arangod/Cluster/Utils/ShardID.h

```diff
@@ -0,0 +1,161 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2023-2023 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "Basics/Exceptions.h"
+#include "Basics/ResultT.h"
+#include "Basics/StringUtils.h"
+#include "Inspection/Status.h"
+
+#include <compare>
+#include <cstdint>
+#include <fmt/format.h>
+#include <velocypack/Value.h>
+
+namespace arangodb {
+
+struct ShardID {
+
+  static ResultT<ShardID> shardIdFromString(std::string const& s) {
+    if (s.empty() || s.at(0) != 's') {
+      return Result{TRI_ERROR_BAD_PARAMETER, "Expected ShardID to start with 's'"};
+    }
+    auto res = basics::StringUtils::try_uint64(s.c_str() + 1, s.length() -1);
+    if (res.fail()) {
+      return std::move(res.result());
+    }
+    return ShardID{res.get()};
+  }
+
+  static ShardID invalidShard() { return ShardID{0}; }
+
+  // Default constructor required for Inspectors.
+  // Note: This shardID is considered Invalid.
+  ShardID() : id(0) {}
+
+  explicit ShardID(uint64_t id) : id(id) {}
+
+  explicit ShardID(std::string_view id) {
+    auto maybeShardID = shardIdFromString(std::string(id));
+    if (maybeShardID.fail()) {
+      THROW_ARANGO_EXCEPTION(maybeShardID.result());
+    }
+    *this = maybeShardID.get();
+  }
+
+
+  ~ShardID() = default;
+  ShardID(ShardID const&) = default;
+  ShardID(ShardID&&) = default;
+  ShardID& operator=(ShardID const&) = default;
+  ShardID& operator=(ShardID&&) = default;
+
+  operator std::string() const {
+    return "s" + std::to_string(id);
+  }
+
+  operator arangodb::velocypack::Value() const {
+    return arangodb::velocypack::Value("s" + std::to_string(id));
```

Comment:
I don't think that this will work safely.
The returned velocypack::Value string points into a temporary that will be invalid when the function scope is left.
This looks like undefined behavior!

Reply:
It indeed is!

Reply:
Deleted by: 8da353b6215

The behavior i required this for is actually covered by the std::string operator, there without undefined behaviour.
Added another Test in the Above commit (FlatHashMap) that uses the same codepath. This test passes witu ASAN/UBSAN now, did not pass with the original VPack::Value operator.